### PR TITLE
Store pipeline state in RC labels

### PR DESCRIPTION
**Pull request by msteffen**
- This eliminates the need for use to EtcdPipelineInfo previous state, which is not always available or accurate
- This makes the PPS master's model closer to that of a kubernetes controller: The `EtcdPipelineInfo` state is the *target*, and the pipeline RC is the current state. The master makes changes to attempt to bring the RC in line with the target (with a few exceptions, e.g. `pipelineInfo.Stopped`)
- It also makes the PPS master logic shorter and more robust in the presence of e.g. missing pipeline RCs: see `step()` in `pipeline_controller.go`

### CODE REVIEW

File: src/server/pkg/collection/collection.go [hunk 1]
```diff
@@ -568,10 +568,6 @@ func (c *readonlyCollection) Watch() (watch.Watcher, error) {
 	return watch.NewWatcher(c.ctx, c.etcdClient, c.prefix, c.prefix, c.template)
 }
 
-func (c *readonlyCollection) WatchWithPrev() (watch.Watcher, error) {
-	return watch.NewWatcherWithPrev(c.ctx, c.etcdClient, c.prefix, c.prefix, c.template)
-}
-
```
Comment (jdoliner) [inline src/server/pkg/collection/collection.go]: As much as I love seeing code being deleted. This seems like a reasonable method to have, even if it's no longer being used.

**Comment by jdoliner** [comment]
This definitely looks like the direction we want the master to go in. It seems like a fairly big change though to some code that's taken a while to get right. Maybe the diff makes it seem larger than it is, but I'm a bit worried about getting this into users hands without more thorough testing than our CI does. How confident do you feel in the code? Maybe I'm just being overly paranoid.

**Comment by msteffen** [comment]
> How confident do you feel in the code? Maybe I'm just being overly paranoid.

I'm of two minds. On the one hand, I feel more confident that this code is right than that the old code wasn't missing some corner case (I'm trying to upload a gist of the two side-by-side but github is hard). OTOH there is a lot of value in code being well-used (though even 1.7.12 had some pretty major changes).

If we want to be conservative, maybe we could ask Allen to roll out 1.7.12, and warn him about the possibility that it could fail pipelines too aggressively? And tell him that if he sees that happening, then he can try the new code

File: src/server/pkg/collection/collection.go (hunk 1 above)
Comment (msteffen) [reply src/server/pkg/collection/collection.go]: TBH, I feel compelled to push on this just a little bit. I think `WithPrev` is generally unsafe, and part of what made the previous version of the PPS master brittle was that you couldn't rely on prevKey being set—if the master restarted, `prevValue` wasn't set for any of the values returned by the watch, which meant that all of the code that handled watch updates had to query kubernetes anyway to see what was there and deal with unexpected cases.

I think that anywhere we use `withPrev` in the future is going to have this problem—I think it's not a generally safe assumption that a watch restarting is a corner case, and as soon as it happens any logic based on the previous key becomes invalid.

If you still disagree I'm happy to leave it

Comment (jdoliner) [reply src/server/pkg/collection/collection.go]: This makes sense, I'm cool with removing it.

**Comment by gabrielgrant** [comment]
Is this something we're still looking to merge in, now that 1.9 has shipped?

**Comment by msteffen** [comment]
Closing this PR, since the corresponding change was made in 1.9.3 and the only remaining 1.7 customer is migrating off

**Pull request closed without merging**